File: sdk/python/bitcoin_bastion_sdk/webhooks.py
from __future__ import annotations

import hashlib
import hmac
import time


def verify_signature(
    *,
    payload: bytes | str,
    secret: str,
    timestamp: int | str,
    signature: str,
    delivery_id: str | None = None,
    event_type: str | None = None,
    tolerance_seconds: int = 300,
    now: int | None = None,
) -> bool:
    observed = int(now if now is not None else time.time())
    ts = int(timestamp)
    if abs(observed - ts) > tolerance_seconds:
        return False
    if not signature.startswith("v1=") or not delivery_id or not event_type:
        return False
    raw = payload if isinstance(payload, bytes) else payload.encode("utf-8")
    message = f"{ts}.{delivery_id}.{event_type}.".encode("utf-8") + raw
    expected = "v1=" + hmac.new(secret.encode("utf-8"), message, hashlib.sha256).hexdigest()
    return hmac.compare_digest(expected, signature)

File: sdk/typescript/src/webhooks.ts
import { createHmac, timingSafeEqual } from "node:crypto";

export interface VerifyWebhookSignatureOptions {
  payload: string | Buffer;
  signature: string;
  timestamp: string | number;
  secret: string;
  deliveryId?: string;
  eventType?: string;
  toleranceSeconds?: number;
}

export function verifyBastionWebhookSignature(options: VerifyWebhookSignatureOptions): boolean {
  const toleranceSeconds = options.toleranceSeconds ?? 300;
  const timestamp = Number(options.timestamp);
  const nowSeconds = Math.floor(Date.now() / 1000);
  if (!Number.isFinite(timestamp) || Math.abs(nowSeconds - timestamp) > toleranceSeconds) return false;
  if (!options.signature.startsWith("v1=") || !options.deliveryId || !options.eventType) return false;
  const rawPayload = Buffer.isBuffer(options.payload) ? options.payload.toString("utf8") : options.payload;
  const expected = `v1=${createHmac("sha256", options.secret).update(`${timestamp}.${options.deliveryId}.${options.eventType}.${rawPayload}`).digest("hex")}`;
  const expectedBuffer = Buffer.from(expected);
  const actualBuffer = Buffer.from(options.signature);
  return expectedBuffer.length === actualBuffer.length && timingSafeEqual(expectedBuffer, actualBuffer);
}
